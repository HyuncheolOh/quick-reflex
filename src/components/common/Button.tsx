import React from 'react';
import {
  TouchableOpacity,
  Text,
  StyleSheet,
  ViewStyle,
  TextStyle,
  ActivityIndicator,
} from 'react-native';
import { COLORS, SPACING, TYPOGRAPHY, BORDER_RADIUS, SHADOWS } from '../../constants';

interface ButtonProps {
  title: string;
  onPress: () => void;
  variant?: 'primary' | 'secondary' | 'danger' | 'ghost';
  size?: 'small' | 'medium' | 'large';
  disabled?: boolean;
  loading?: boolean;
  style?: ViewStyle;
  textStyle?: TextStyle;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  onPress,
  variant = 'primary',
  size = 'medium',
  disabled = false,
  loading = false,
  style,
  textStyle,
}) => {
  const buttonStyle = [
    styles.base,
    styles[`${variant}Container`],
    styles[`${size}Container`],
    disabled && styles.disabledContainer,
    style,
  ];

  const titleStyle = [
    styles.baseText,
    styles[`${variant}Text`],
    styles[`${size}Text`],
    disabled && styles.disabledText,
    textStyle,
  ];

  return (
    <TouchableOpacity
      style={buttonStyle}
      onPress={onPress}
      disabled={disabled || loading}
      activeOpacity={0.8}
    >
      {loading ? (
        <ActivityIndicator
          size="small"
          color={variant === 'ghost' ? COLORS.PRIMARY : COLORS.TEXT_PRIMARY}
        />
      ) : (
        <Text style={titleStyle}>{title}</Text>
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  base: {
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: BORDER_RADIUS.LG,
    ...SHADOWS.SMALL,
  },
  
  // Container variants
  primaryContainer: {
    backgroundColor: COLORS.PRIMARY,
  },
  secondaryContainer: {
    backgroundColor: COLORS.SECONDARY,
  },
  dangerContainer: {
    backgroundColor: COLORS.ERROR,
  },
  ghostContainer: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: COLORS.PRIMARY,
    ...SHADOWS.SMALL,
  },
  
  // Container sizes
  smallContainer: {
    paddingHorizontal: SPACING.MD,
    paddingVertical: SPACING.SM,
    minHeight: 36,
  },
  mediumContainer: {
    paddingHorizontal: SPACING.LG,
    paddingVertical: SPACING.MD,
    minHeight: 44,
  },
  largeContainer: {
    paddingHorizontal: SPACING.XL,
    paddingVertical: SPACING.LG,
    minHeight: 52,
  },
  
  // Disabled state
  disabledContainer: {
    opacity: 0.5,
  },
  
  // Base text
  baseText: {
    textAlign: 'center',
    fontWeight: TYPOGRAPHY.FONT_WEIGHT.SEMIBOLD,
  },
  
  // Text variants
  primaryText: {
    color: COLORS.TEXT_PRIMARY,
  },
  secondaryText: {
    color: COLORS.TEXT_PRIMARY,
  },
  dangerText: {
    color: COLORS.TEXT_PRIMARY,
  },
  ghostText: {
    color: COLORS.PRIMARY,
  },
  
  // Text sizes
  smallText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.SM,
  },
  mediumText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.MD,
  },
  largeText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
  },
  
  // Disabled text
  disabledText: {
    opacity: 0.7,
  },
});