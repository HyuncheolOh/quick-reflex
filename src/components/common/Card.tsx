import React from 'react';
import {
  View,
  StyleSheet,
  ViewStyle,
  TouchableOpacity,
  TouchableOpacityProps,
} from 'react-native';
import { COLORS, SPACING, BORDER_RADIUS, SHADOWS } from '../../constants';

interface CardProps extends TouchableOpacityProps {
  children: React.ReactNode;
  style?: ViewStyle;
  padding?: keyof typeof SPACING;
  touchable?: boolean;
  variant?: 'default' | 'elevated' | 'outlined';
}

export const Card: React.FC<CardProps> = ({
  children,
  style,
  padding = 'MD',
  touchable = false,
  variant = 'default',
  ...touchableProps
}) => {
  const cardStyle = [
    styles.base,
    styles[variant],
    { padding: SPACING[padding] },
    style,
  ];

  if (touchable) {
    return (
      <TouchableOpacity style={cardStyle} activeOpacity={0.8} {...touchableProps}>
        {children}
      </TouchableOpacity>
    );
  }

  return <View style={cardStyle}>{children}</View>;
};

const styles = StyleSheet.create({
  base: {
    borderRadius: BORDER_RADIUS.LG,
  },
  
  default: {
    backgroundColor: COLORS.CARD,
  },
  
  elevated: {
    backgroundColor: COLORS.CARD,
    ...SHADOWS.MEDIUM,
  },
  
  outlined: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: COLORS.TEXT_TERTIARY,
  },
});