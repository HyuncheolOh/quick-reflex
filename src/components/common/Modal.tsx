import React from 'react';
import {
  Modal as RNModal,
  View,
  StyleSheet,
  Dimensions,
  TouchableOpacity,
  Text,
} from 'react-native';
import { COLORS, SPACING, TYPOGRAPHY, BORDER_RADIUS } from '../../constants';
import { Button } from './Button';

const { width, height } = Dimensions.get('window');

interface ModalProps {
  visible: boolean;
  onClose: () => void;
  title?: string;
  children: React.ReactNode;
  showCloseButton?: boolean;
  animationType?: 'slide' | 'fade' | 'none';
}

export const Modal: React.FC<ModalProps> = ({
  visible,
  onClose,
  title,
  children,
  showCloseButton = true,
  animationType = 'fade',
}) => {
  return (
    <RNModal
      visible={visible}
      transparent
      animationType={animationType}
      onRequestClose={onClose}
    >
      <View style={styles.overlay}>
        <TouchableOpacity
          style={styles.backdrop}
          activeOpacity={1}
          onPress={onClose}
        />
        
        <View style={styles.container}>
          {title && (
            <View style={styles.header}>
              <Text style={styles.title}>{title}</Text>
              {showCloseButton && (
                <TouchableOpacity onPress={onClose} style={styles.closeButton}>
                  <Text style={styles.closeButtonText}>✕</Text>
                </TouchableOpacity>
              )}
            </View>
          )}
          
          <View style={styles.content}>
            {children}
          </View>
        </View>
      </View>
    </RNModal>
  );
};

export const ConfirmModal: React.FC<{
  visible: boolean;
  onClose: () => void;
  onConfirm: () => void;
  title: string;
  message: string;
  confirmText?: string;
  cancelText?: string;
  variant?: 'danger' | 'primary';
}> = ({
  visible,
  onClose,
  onConfirm,
  title,
  message,
  confirmText = '확인',
  cancelText = '취소',
  variant = 'primary',
}) => {
  return (
    <Modal visible={visible} onClose={onClose} title={title} showCloseButton={false}>
      <Text style={styles.message}>{message}</Text>
      
      <View style={styles.buttonContainer}>
        <Button
          title={cancelText}
          onPress={onClose}
          variant="ghost"
          style={styles.button}
        />
        <Button
          title={confirmText}
          onPress={() => {
            onConfirm();
            onClose();
          }}
          variant={variant === 'danger' ? 'danger' : 'primary'}
          style={styles.button}
        />
      </View>
    </Modal>
  );
};

const styles = StyleSheet.create({
  overlay: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
  },
  
  container: {
    backgroundColor: COLORS.SURFACE,
    borderRadius: BORDER_RADIUS.XL,
    maxWidth: width - SPACING.XL * 2,
    maxHeight: height - SPACING.XXL * 2,
    minWidth: 280,
  },
  
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    paddingHorizontal: SPACING.LG,
    paddingTop: SPACING.LG,
    paddingBottom: SPACING.MD,
    borderBottomWidth: 1,
    borderBottomColor: COLORS.TEXT_TERTIARY,
  },
  
  title: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
    fontWeight: TYPOGRAPHY.FONT_WEIGHT.BOLD,
    color: COLORS.TEXT_PRIMARY,
    flex: 1,
  },
  
  closeButton: {
    padding: SPACING.SM,
    marginLeft: SPACING.MD,
  },
  
  closeButtonText: {
    fontSize: TYPOGRAPHY.FONT_SIZE.LG,
    color: COLORS.TEXT_SECONDARY,
  },
  
  content: {
    padding: SPACING.LG,
  },
  
  message: {
    fontSize: TYPOGRAPHY.FONT_SIZE.MD,
    color: COLORS.TEXT_PRIMARY,
    textAlign: 'center',
    marginBottom: SPACING.LG,
    lineHeight: 22,
  },
  
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    gap: SPACING.MD,
  },
  
  button: {
    flex: 1,
  },
});