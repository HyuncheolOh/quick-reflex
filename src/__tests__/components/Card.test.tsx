import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { Text } from 'react-native';
import { Card } from '../../components/common/Card';

// Mock useThemedColors hook
jest.mock('../../hooks', () => ({
  useThemedColors: () => ({
    CARD: '#FFFFFF',
    TEXT_TERTIARY: '#8E8E93',
  }),
}));

describe('Card', () => {
  const TestContent = () => <Text>Card Content</Text>;

  describe('Basic Rendering', () => {
    it('should render children content', () => {
      const { getByText } = render(
        <Card>
          <TestContent />
        </Card>
      );
      
      expect(getByText('Card Content')).toBeTruthy();
    });

    it('should render as View by default', () => {
      const { getByTestId } = render(
        <Card testID="card">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.type).toBe('View');
    });

    it('should render as TouchableOpacity when touchable is true', () => {
      const { getByTestId } = render(
        <Card testID="card" touchable>
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.type).toBe('TouchableOpacity');
    });
  });

  describe('Variants', () => {
    it('should render default variant', () => {
      const { getByTestId } = render(
        <Card testID="card">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: '#FFFFFF',
          }),
        ])
      );
    });

    it('should render elevated variant with shadow', () => {
      const { getByTestId } = render(
        <Card testID="card" variant="elevated">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: '#FFFFFF',
            shadowColor: '#000',
          }),
        ])
      );
    });

    it('should render outlined variant', () => {
      const { getByTestId } = render(
        <Card testID="card" variant="outlined">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            backgroundColor: 'transparent',
            borderWidth: 1,
            borderColor: '#8E8E93',
          }),
        ])
      );
    });
  });

  describe('Padding', () => {
    it('should apply medium padding by default', () => {
      const { getByTestId } = render(
        <Card testID="card">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            padding: 16, // MD spacing
          }),
        ])
      );
    });

    it('should apply small padding', () => {
      const { getByTestId } = render(
        <Card testID="card" padding="SM">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            padding: 8, // SM spacing
          }),
        ])
      );
    });

    it('should apply large padding', () => {
      const { getByTestId } = render(
        <Card testID="card" padding="LG">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            padding: 24, // LG spacing
          }),
        ])
      );
    });

    it('should apply extra large padding', () => {
      const { getByTestId } = render(
        <Card testID="card" padding="XL">
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            padding: 32, // XL spacing
          }),
        ])
      );
    });
  });

  describe('Touchable Behavior', () => {
    it('should call onPress when touchable card is pressed', () => {
      const mockOnPress = jest.fn();
      const { getByTestId } = render(
        <Card testID="card" touchable onPress={mockOnPress}>
          <TestContent />
        </Card>
      );
      
      fireEvent.press(getByTestId('card'));
      expect(mockOnPress).toHaveBeenCalledTimes(1);
    });

    it('should not be pressable when touchable is false', () => {
      const mockOnPress = jest.fn();
      const { getByTestId } = render(
        <Card testID="card" touchable={false} onPress={mockOnPress}>
          <TestContent />
        </Card>
      );
      
      // Should not be able to press a non-touchable card
      const card = getByTestId('card');
      expect(card.type).toBe('View');
    });

    it('should pass through TouchableOpacity props', () => {
      const mockOnPress = jest.fn();
      const mockOnLongPress = jest.fn();
      
      const { getByTestId } = render(
        <Card 
          testID="card" 
          touchable 
          onPress={mockOnPress}
          onLongPress={mockOnLongPress}
          activeOpacity={0.5}
        >
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.onPress).toBe(mockOnPress);
      expect(card.props.onLongPress).toBe(mockOnLongPress);
      expect(card.props.activeOpacity).toBe(0.5);
    });
  });

  describe('Custom Styles', () => {
    it('should apply custom styles', () => {
      const customStyle = {
        marginTop: 20,
        borderRadius: 8,
      };
      
      const { getByTestId } = render(
        <Card testID="card" style={customStyle}>
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([customStyle])
      );
    });

    it('should merge custom styles with base styles', () => {
      const customStyle = { backgroundColor: 'red' };
      
      const { getByTestId } = render(
        <Card testID="card" style={customStyle}>
          <TestContent />
        </Card>
      );
      
      const card = getByTestId('card');
      expect(card.props.style).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            borderRadius: 12, // Base border radius
          }),
          customStyle,
        ])
      );
    });
  });


  describe('Content Rendering', () => {
    it('should render multiple children', () => {
      const { getByText } = render(
        <Card>
          <Text>First Child</Text>
          <Text>Second Child</Text>
        </Card>
      );
      
      expect(getByText('First Child')).toBeTruthy();
      expect(getByText('Second Child')).toBeTruthy();
    });

    it('should render complex nested content', () => {
      const { getByText, getByTestId } = render(
        <Card>
          <TestContent />
          <Card testID="nested-card" variant="outlined">
            <Text>Nested Content</Text>
          </Card>
        </Card>
      );
      
      expect(getByText('Card Content')).toBeTruthy();
      expect(getByText('Nested Content')).toBeTruthy();
      expect(getByTestId('nested-card')).toBeTruthy();
    });
  });

  describe('Snapshot Tests', () => {
    it('should match snapshot for default card', () => {
      const { toJSON } = render(
        <Card>
          <TestContent />
        </Card>
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for touchable card', () => {
      const { toJSON } = render(
        <Card touchable>
          <TestContent />
        </Card>
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for elevated card', () => {
      const { toJSON } = render(
        <Card variant="elevated">
          <TestContent />
        </Card>
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot for outlined card', () => {
      const { toJSON } = render(
        <Card variant="outlined">
          <TestContent />
        </Card>
      );
      expect(toJSON()).toMatchSnapshot();
    });
  });
});