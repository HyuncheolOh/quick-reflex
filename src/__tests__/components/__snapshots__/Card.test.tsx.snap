// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Card Snapshot Tests should match snapshot for default card 1`] = `
<View
  style={
    [
      {
        "borderRadius": 12,
      },
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "padding": 16,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Snapshot Tests should match snapshot for elevated card 1`] = `
<View
  style={
    [
      {
        "borderRadius": 12,
      },
      {
        "backgroundColor": "#FFFFFF",
        "elevation": 5,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.25,
        "shadowRadius": 3.84,
      },
      {
        "padding": 16,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Snapshot Tests should match snapshot for outlined card 1`] = `
<View
  style={
    [
      {
        "borderRadius": 12,
      },
      {
        "backgroundColor": "transparent",
        "borderColor": "#8E8E93",
        "borderWidth": 1,
      },
      {
        "padding": 16,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</View>
`;

exports[`Card Snapshot Tests should match snapshot for touchable card 1`] = `
<TouchableOpacity
  activeOpacity={0.8}
  style={
    [
      {
        "borderRadius": 12,
      },
      {
        "backgroundColor": "#FFFFFF",
      },
      {
        "padding": 16,
      },
      undefined,
    ]
  }
>
  <Text>
    Card Content
  </Text>
</TouchableOpacity>
`;
